### PR TITLE
IsInf ReduceSum transform

### PR DESCRIPTION
**Pull request by ashbhandare**
This transform changes the following subgraph pattern:
Cast --> IsInf --> Cast --> ReduceSum --> Greater
to
IsAllFinite --> Not

This PR also adds attributes to the contrib op IsAllFinite to check for only inf or nan.

This optimization benefits by:
1) Fused cuda kernel being used for IsAllFinite instead of doing IsInf on CPU(no CUDA kernel implemented), and ReduceSum + Greater
2) avoids Casts

### CODE REVIEW

File: onnxruntime/test/testdata/transform/fusion/isinf_reducesum.py [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+kwargs={}
+kwargs['opset_imports'] = opsets
+
+def GenerateModel(model_name):
+    nodes = [  # subgraph
+        helper.make_node("Cast", ["A"], ["cast1"], "cast_1", to=11),
+
+        helper.make_node("IsInf", ["cast1"], ["IsInf_out"], "is_inf"),
+
+        helper.make_node("Cast", ["IsInf_out"], ["cast2"], "cast_2", to=7),
```
Comment (ytaous) [inline onnxruntime/test/testdata/transform/fusion/isinf_reducesum.py:26]: would these two casts always be there?

Comment (ytaous) [reply onnxruntime/test/testdata/transform/fusion/isinf_reducesum.py:26]: saw your comments in the code

---
In reply to: [604599029](https://github.com/microsoft/onnxruntime/pull/7188#discussion_r604599029) [](ancestors = 604599029)

File: onnxruntime/core/graph/graph_utils.cc [hunk 2]
```diff
@@ -715,6 +715,17 @@ void FinalizeNodeFusion(Graph& graph, const std::vector<std::reference_wrapper<N
   }
 }
 
+void FinalizeNodeFusion(Graph& graph, const std::vector<std::reference_wrapper<Node>>& nodes, Node& replacement_node_start,
+Node& replacement_node_end) {
```
Comment (ytaous) [inline onnxruntime/core/graph/graph_utils.cc:719]: nit: format

**Comment by ytaous** [comment]
Add description pls.

File: orttraining/orttraining/training_ops/cuda/math/isfinite.cuh [hunk 3]
```diff
@@ -8,12 +8,23 @@
 namespace onnxruntime {
 namespace cuda {
 
-template<typename T>
+template <typename T>
 __device__ __forceinline__ bool _IsFiniteScalar(const T value) {
```
Comment (yuslepukhin) [inline orttraining/orttraining/training_ops/cuda/math/isfinite.cuh:12]: >_ [](start = 32, length = 1)

Generally, names starting with underscore are reserved names in C++. This was an unforntunate pattern started by someone.

File: orttraining/orttraining/training_ops/cuda/math/isfinite.h [hunk 4]
```diff
@@ -25,14 +25,27 @@ template <typename TSrc>
...
   Status ComputeInternal(OpKernelContext* context) const override;
+
+  private:
+  bool isinf_only_, isnan_only_;
 };
 
 template <typename T>
 struct IsAllFiniteFunctor {
-  void operator()(cudaStream_t stream, ChunkGroup<1> chunks, bool* output); 
+  void operator()(cudaStream_t stream, ChunkGroup<1> chunks, bool* output, bool isinf_only, bool isnan_only);
```
Comment (yuslepukhin) [inline orttraining/orttraining/training_ops/cuda/math/isfinite.h:48]: >); [](start = 107, length = 2)

nit: const

**Review by ytaous — DISMISSED** [review]

**Comment by SherlockNoMad** [comment]
Please invite Wei-Sheng to review, he is the original author of IsAllFinite.

File: orttraining/orttraining/training_ops/cuda/math/isfinite.cuh (hunk 3 above)
Comment (wschin) [reply orttraining/orttraining/training_ops/cuda/math/isfinite.cuh]: Sorry. I was the someone... lol

---
In reply to: [606355806](https://github.com/microsoft/onnxruntime/pull/7188#discussion_r606355806) [](ancestors = 606355806)

File: onnxruntime/contrib_ops/cuda/math/isfinite.cu [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+template <typename T>
+void IsAllFiniteFunctor<T>::operator()(cudaStream_t stream, ChunkGroup<1> chunks, bool* output, const bool isinf_only, const bool isnan_only) {
+  const int block_count = chunks.chunk_count;
+  const int thread_count = ChunkGroup<1>::thread_count_per_block;
+  IsAllFiniteMultiTensorImpl<T><<<block_count, thread_count, 0, stream>>>(chunks, output, isinf_only, isnan_only);
+}
+
+#define INSTANTIATE_ISALLFINITE_FUNCTOR(T) \
+  template void IsAllFiniteFunctor<T>::operator()(cudaStream_t stream, ChunkGroup<1> chunks, bool* output, const bool isinf_only, const bool isnan_only);
+
```
Comment (wschin) [inline onnxruntime/contrib_ops/cuda/math/isfinite.cu:44]: I have a feeling that some lines are too long. Could you wrap them? Thanks.

File: onnxruntime/contrib_ops/cuda/math/isfinite.cuh [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+__device__ __forceinline__ bool IsFiniteScalar(const half value) {
+#if __CUDA_ARCH__ >= 530 || !defined(__CUDA_ARCH__)
+  return !__hisinf(value) && !__hisnan(value);
+#else
+  return isfinite(float(value));
+#endif
+}
+
+template <>
+__device__ __forceinline__ bool IsFiniteScalar(const half value, const bool isinf_only, const bool isnan_only) {
```
Comment (wschin) [inline onnxruntime/contrib_ops/cuda/math/isfinite.cuh:37]: >const bool isinf_only, const bool isnan_only [](start = 65, length = 44)

Let's make them template parameters.

File: onnxruntime/contrib_ops/cuda/math/isfinite.cuh [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+    return isfinite(float(value));
+  }
+#endif
+}
+
+#if CUDA_VERSION >= 11000 && (__CUDA_ARCH__ >= 800 || !defined(__CUDA_ARCH__))
+template <>
+__device__ __forceinline__ bool IsFiniteScalar(const nv_bfloat16 value) {
+  return isfinite(float(value));
+}
```
Comment (wschin) [inline onnxruntime/contrib_ops/cuda/math/isfinite.cuh:61]: Is it possible to use NV's intrinsic like __hisinf above? Maybe this ref is helpful. [https://docs.nvidia.com/cuda/cuda-math-api/group__CUDA__MATH____BFLOAT16__COMPARISON.html#group__CUDA__MATH____BFLOAT16__COMPARISON](https://docs.nvidia.com/cuda/cuda-math-api/group__CUDA__MATH____BFLOAT16__COMPARISON.html#group__CUDA__MATH____BFLOAT16__COMPARISON)

File: onnxruntime/contrib_ops/cuda/math/isfinite.cuh [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+
+template <>
+__device__ __forceinline__ bool IsFiniteScalar(const nv_bfloat16 value, const bool isinf_only, const bool isnan_only) {
```
Comment (wschin) [inline onnxruntime/contrib_ops/cuda/math/isfinite.cuh:64]: >const bool isinf_only, const bool isnan_only [](start = 72, length = 44)

Similar to another comment, they can be template paramters.

File: onnxruntime/core/graph/graph_utils.h [hunk 9]
```diff
@@ -213,6 +213,19 @@ void FinalizeNodeFusion(Graph& graph, Node& first_node, Node& second_node);
...
+
+    Conceptually multiple nodes are being combined, and post-fusion will produce output/s with the same names
+    as the last node in 'nodes', and be connected to the same downstream nodes.
+
+    The input edges to the first node in 'nodes' will be moved to replacement_node. No other input edges are moved.
+    The output definitions and edges from the last node in 'nodes' will be moved to replacement_node.
+    All nodes in 'nodes' will be removed.
+*/
+void FinalizeNodeFusion(Graph& graph, const std::vector<std::reference_wrapper<Node>>& nodes, Node& replacement_node_start,
+Node& replacement_node_end);
```
Comment (wschin) [inline onnxruntime/core/graph/graph_utils.h:227]: Missing spaces. If the argument list is long, we can wrap it to multiple lines.

**Comment by wschin** [comment]
>   MoveAllNodeInputEdges(graph, nodes.front(), replacement_node);

Let's rewrite this function using
```
FinalizeNodeFusion(graph, nodes, replacement_node, replacement_node)
```

---
Refers to: onnxruntime/core/graph/graph_utils.cc:709 in 2ef04ca. [](commit_id = 2ef04ca7d322622c132a7a1a12746a28b3282be5, deletion_comment = False)

File: onnxruntime/core/graph/graph_utils.cc [hunk 10]
```diff
@@ -715,6 +715,17 @@ void FinalizeNodeFusion(Graph& graph, const std::vector<std::reference_wrapper<N
...
+  MoveAllNodeInputEdges(graph, nodes.front(), replacement_node_start);
+  MoveAllNodeOutputs(graph, nodes.back(), replacement_node_end);
+
+  for (Node& node : nodes) {
+    RemoveNodeOutputEdges(graph, node);
```
Comment (wschin) [inline onnxruntime/core/graph/graph_utils.cc:724]: >RemoveNodeOutputEdges [](start = 4, length = 21)

Would this removal cause missing input to other nodes outside `nodes`?

File: onnxruntime/core/optimizer/isinf_reducesum_fusion.cc [hunk 11]
```diff
@@ -0,0 +1,129 @@
...
+  const auto& node_topology_list = graph_viewer.GetNodesInTopologicalOrder();
+  std::vector<std::reference_wrapper<Node>> nodes_to_remove;
+  for (auto node_index : node_topology_list) {
+    nodes_to_remove.clear();
+    auto* node_ptr = graph.GetNode(node_index);
+    if (!node_ptr)
+      continue;  // node was removed
+
+    auto& isinf_node = *node_ptr;
+
```
Comment (wschin) [inline onnxruntime/core/optimizer/isinf_reducesum_fusion.cc:33]: `ORT_ENFORCE(node_ptr->OpType() == "IsInf"` maybe?

File: onnxruntime/core/optimizer/isinf_reducesum_fusion.cc [hunk 12]
```diff
@@ -0,0 +1,129 @@
...
+    }
+
+    std::vector<NodeArg*> input_defs = isinf_node.MutableInputDefs();
+    // see if there is a Cast before IsInf
+    // This will happen if input type is FP16 but IsInf doesnt support fp16, so it will be cast to float/double
+    // This Cast can be skipped as we are replacing the subgraph with IsAllFinite, which supports FP16
+    auto cast1_node_iter = isinf_node.InputNodesBegin();
+    if (cast1_node_iter != isinf_node.InputNodesEnd() &&
+        graph_utils::IsSupportedOptypeVersionAndDomain(*cast1_node_iter, "Cast", {9, 13}) &&
+        cast1_node_iter->GetOutputEdgesCount() == 1) {
```
Comment (wschin) [inline onnxruntime/core/optimizer/isinf_reducesum_fusion.cc:49]: >cast1_node_iter->GetOutputEdgesCount() == 1 [](start = 8, length = 43)

Great. So we check if this Cast doesn't have other consumer.

File: onnxruntime/core/optimizer/isinf_reducesum_fusion.cc [hunk 13]
```diff
@@ -0,0 +1,129 @@
...
+      Node& cast1_node = *graph.GetNode(cast1_node_iter->Index());
+      auto cast1_input_defs = cast1_node.MutableInputDefs();
+      auto cast1_input_type = cast1_input_defs[0] ? cast1_input_defs[0]->Type() : nullptr;
+      // remove cast only if the input type is float16
+      if (cast1_input_type != nullptr && (*cast1_input_type) == "tensor(float16)") {
+        input_defs = cast1_input_defs;
+        nodes_to_remove.push_back(cast1_node);
+      }
+    }
+    nodes_to_remove.push_back(isinf_node);
```
Comment (wschin) [inline onnxruntime/core/optimizer/isinf_reducesum_fusion.cc:60]: >nodes_to_remove.push_back(isinf_node); [](start = 4, length = 38)

What if the first "Cast" has other cusomers? Would that fail this optimization?

File: onnxruntime/core/optimizer/isinf_reducesum_fusion.cc (hunk 12 above)
Comment (ashbhandare) [reply onnxruntime/core/optimizer/isinf_reducesum_fusion.cc:49]: that's correct

File: onnxruntime/core/optimizer/isinf_reducesum_fusion.cc (hunk 13 above)
Comment (ashbhandare) [reply onnxruntime/core/optimizer/isinf_reducesum_fusion.cc:60]: we check for it

File: onnxruntime/core/optimizer/isinf_reducesum_fusion.cc (hunk 11 above)
Comment (ashbhandare) [reply onnxruntime/core/optimizer/isinf_reducesum_fusion.cc:33]: we check for it in the next few lines

File: onnxruntime/contrib_ops/cuda/math/isfinite.cuh (hunk 6 above)
Comment (ashbhandare) [reply onnxruntime/contrib_ops/cuda/math/isfinite.cuh:37]: reconfigured to templatize

File: onnxruntime/core/graph/graph_utils.cc (hunk 10 above)
Comment (ashbhandare) [reply onnxruntime/core/graph/graph_utils.cc:724]: If any of the nodes in `nodes` have consumers that are not being removed, the fusions should not be triggered. The fusion implementations should already check for this case

**Comment by ashbhandare** [comment]
> > MoveAllNodeInputEdges(graph, nodes.front(), replacement_node);
> 
> Let's rewrite this function using
> 
> ```
> FinalizeNodeFusion(graph, nodes, replacement_node, replacement_node)
> ```
> 
> Refers to: onnxruntime/core/graph/graph_utils.cc:709 in [2ef04ca](https://github.com/microsoft/onnxruntime/commit/2ef04ca7d322622c132a7a1a12746a28b3282be5). [](commit_id = [2ef04ca](https://github.com/microsoft/onnxruntime/commit/2ef04ca7d322622c132a7a1a12746a28b3282be5), deletion_comment = False)

Rewriting as discussed offline

File: onnxruntime/test/optimizer/graph_transform_test.cc [hunk 14]
```diff
@@ -3818,6 +3819,27 @@ TEST_F(GraphTransformationTests, MatMulScaleFusionUnsupportedInputType) {
       },
       {kCpuExecutionProvider});
 }
+
+#if defined(USE_CUDA)
+TEST_F(GraphTransformationTests, IsInfReduceSum_Test) {
```
Comment (weixingzhang) [inline onnxruntime/test/optimizer/graph_transform_test.cc:3824]: Could you please enable it for ROCm EP as well?

**Review by wschin — APPROVED** [review]
LTGM. Thanks.

**Pull request merged**